crates/editor/src/display_map.rs

```diff
@@ -1514,6 +1540,228 @@ impl ToDisplayPoint for Anchor {
+                    &buffer,
+                );
+
+                Some(MultibufferSemanticToken {
+                    range: start_offset..end_offset,
+                    style: stylizer.convert(
+                        cx.theme().syntax(),
+                        token.token_type,
+                        token.token_modifiers,
+                    )?,
+                    lsp_type: token.token_type,
+                    lsp_modifiers: token.token_modifiers,
+                })
+            })
+            .collect::<Vec<_>>();
+
+        // These should be sorted, but we rely on it for binary searching, so let's be sure.
+        tokens.sort_by_key(|token| token.range.start);
+
+        Some(SemanticTokenView {
+            tokens,
+            version: buffer.version(),
+        })
+    }
+
+    pub fn tokens_in_range(&self, range: Range<usize>) -> &[MultibufferSemanticToken] {
+        let start = self
+            .tokens
+            .binary_search_by_key(&range.start, |token| token.range.start)
+            .unwrap_or_else(|next_ix| next_ix);
+
+        let end = self
+            .tokens
+            .binary_search_by_key(&range.end, |token| token.range.start)
+            .unwrap_or_else(|next_ix| next_ix);
+
+        &self.tokens[start..end]
+    }
+}
+
+fn point_offset_to_offsets(
+    point: text::PointUtf16,
+    length: text::OffsetUtf16,
+    buffer: &text::Buffer,
+) -> (usize, usize) {
+    let start = buffer.as_rope().point_utf16_to_offset(point);
+    let start_offset = buffer.as_rope().offset_to_offset_utf16(start);
+    let end_offset = start_offset + length;
+    let end = buffer.as_rope().offset_utf16_to_offset(end_offset);
+
+    (start, end)
+}
+
+struct SemanticTokenStylizer<'a> {
+    token_types: Vec<&'a str>,
+    modifier_mask: HashMap<&'a str, u32>,
+}
+
+impl<'a> SemanticTokenStylizer<'a> {
+    pub fn new(legend: &'a lsp::SemanticTokensLegend) -> Self {
+        let token_types = legend.token_types.iter().map(|s| s.as_str()).collect();
+        let modifier_mask = legend
+            .token_modifiers
+            .iter()
+            .enumerate()
+            .map(|(i, modifier)| (modifier.as_str(), 1 << i))
+            .collect();
+        SemanticTokenStylizer {
+            token_types,
+            modifier_mask,
+        }
+    }
+
+    pub fn token_type(&self, token_type: u32) -> Option<&'a str> {
+        self.token_types.get(token_type as usize).copied()
+    }
+
+    pub fn has_modifier(&self, token_modifiers: u32, modifier: &str) -> bool {
+        let Some(mask) = self.modifier_mask.get(modifier) else {
+            return false;
+        };
+        (token_modifiers & mask) != 0
+    }
+
+    pub fn convert(
+        &self,
+        theme: &'a SyntaxTheme,
+        token_type: u32,
+        modifiers: u32,
+    ) -> Option<HighlightStyle> {
+        let has_modifier = |modifier| self.has_modifier(modifiers, modifier);
+
+        // See the VSCode docs [1] and the LSP Spec [2]
```

Comment:
This all seems misplaced: `display_map.rs` deals with coordinate transformations between different layers (inlays, folds, tabs) from text buffer into a rendered editor.

Adding LSP into the mix this way seems odd, when we have `LspStore` and related modules.
Method's parameters do not need anything editor-specific, it seems?